ext/ddtrace_profiling_native_extension/private_vm_api_access.c

```diff
@@ -767,3 +767,40 @@ check_method_entry(VALUE obj, int can_be_svar)
   // they're always on the main Ractor
   bool ddtrace_rb_ractor_main_p(void) { return true; }
 #endif // NO_RACTORS
+
+// This is a tweaked and inlined version of
+// threadptr_invoke_proc_location + rb_proc_location + iseq_location .
+//
+// It's useful to have here because not all of the methods above are accessible to extensions + to avoid the
+// array allocation that iseq_location did to contain its return value.
+static const rb_iseq_t *maybe_thread_invoke_proc_iseq(VALUE thread_value) {
+  rb_thread_t *thread = thread_struct_from_object(thread_value);
+
+  #ifndef NO_THREAD_INVOKE_ARG // Ruby 2.6+
+    if (thread->invoke_type != thread_invoke_type_proc) return NULL;
+
+    VALUE proc = thread->invoke_arg.proc.proc;
+  #else
+    if (thread->first_func || !thread->first_proc) return NULL;
+
+    VALUE proc = thread->first_proc;
+  #endif
+
+  const rb_iseq_t *iseq = rb_proc_get_iseq(proc, 0);
+  if (iseq == NULL) return NULL;
+
+  rb_iseq_check(iseq);
+  return iseq;
+}
+
+VALUE invoke_file_location_for(VALUE thread) {
+  const rb_iseq_t *iseq = maybe_thread_invoke_proc_iseq(thread);
+
+  return iseq != NULL ? rb_iseq_path(iseq) : Qnil;
+}
+
+VALUE invoke_line_location_for(VALUE thread) {
+  const rb_iseq_t *iseq = maybe_thread_invoke_proc_iseq(thread);
+
+  return iseq != NULL ? rb_iseq_first_lineno(iseq) : Qnil;
+}
```

Comment:
```suggestion
VALUE invoke_file_location_for(const rb_iseq_t *iseq) {
  return iseq != NULL ? rb_iseq_path(iseq) : Qnil;
}

VALUE invoke_line_location_for(const rb_iseq_t *iseq) {
  return iseq != NULL ? rb_iseq_first_lineno(iseq) : Qnil;
}
```

Do you think it makes sense to only invoke `maybe_thread_invoke_proc_iseq` once and reuse the value when calling `invoke_file_location_for` and `invoke_line_location_for`?

I understand it exposes `maybe_thread_invoke_proc_iseq` outside of this file, but all the code in question here is internal, and `maybe_thread_invoke_proc_iseq` performs quite a few operations that are being done twice today with no added value.

In either case, the overhead is probably small, but I wanted to chime in because all work here is internal anyway.

Reply:
Agreed, on top of doing the lookup twice, having the two methods made it kinda awkward to the caller as well. I've merged the two functions into one in 87bd64e1f0eea09538ddaf73028bc6104ec73fee, let me know what you think.